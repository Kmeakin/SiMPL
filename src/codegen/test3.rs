use super::{closure::convert, llvm3::Compiler};
use crate::{hir::Expr, types::infer_and_apply};
use inkwell::{context::Context, OptimizationLevel};
use insta::assert_snapshot;
use std::str::FromStr;

#[track_caller]
fn test_compile(src: &str) {
    let expr = Expr::from_str(src).unwrap();
    let expr = infer_and_apply(&expr);
    let cexpr = convert(expr);

    let ctx = Context::create();
    let builder = ctx.create_builder();
    let module = ctx.create_module("test_compile");

    let compiler = Compiler {
        llvm: &ctx,
        module,
        builder,
    };

    let module = compiler.compile_toplevel(&cexpr);

    match module.verify() {
        Ok(()) => {}
        Err(s) => {
            println!("{}\n", module.print_to_string().to_string());
            eprintln!("{}", s.to_string());
            panic!()
        }
    }

    assert_snapshot!(module.print_to_string().to_string());
}

#[track_caller]
fn test_compile_and_execute<T: std::fmt::Debug + PartialEq>(src: &str, expected: T) {
    let expr = Expr::from_str(src).unwrap();
    let expr = infer_and_apply(&expr);
    let cexpr = convert(expr);

    let ctx = Context::create();
    let builder = ctx.create_builder();
    let module = ctx.create_module("test_compile");

    let compiler = Compiler {
        llvm: &ctx,
        module,
        builder,
    };

    let module = compiler.compile_toplevel(&cexpr);

    match module.verify() {
        Ok(()) => {}
        Err(s) => {
            println!("{}\n", module.print_to_string().to_string());
            eprintln!("{}", s.to_string());
            panic!()
        }
    }

    assert_snapshot!(module.print_to_string().to_string());

    let exec_engine = module
        .create_jit_execution_engine(OptimizationLevel::None)
        .unwrap();
    let f = unsafe { exec_engine.get_function::<unsafe extern "C" fn() -> T>("toplevel") }.unwrap();
    assert_eq!(unsafe { f.call() }, expected)
}

#[test]
fn compile_lit() {
    test_compile_and_execute("0", 0);
    test_compile_and_execute("1", 1);
    test_compile_and_execute("true", true);
    test_compile_and_execute("false", false);
    test_compile_and_execute("0.0", 0.0);
    test_compile_and_execute("4.5", 4.5);
}

#[test]
fn compile_vars() {
    test_compile_and_execute("let x = 5 in x", 5);
}

#[test]
fn compile_if() {
    test_compile_and_execute("if true then 5 else 10", 5);
    test_compile_and_execute("if false then 5 else 10", 10);
    test_compile_and_execute(r"let f = \b -> if b then 5 else 10 in f true", 5);
    test_compile_and_execute(r"let f = \b -> if b then 5 else 10 in f false", 10);
}

#[test]
fn compile_lambda() {
    // test_compile(r"\x: Int -> x");
    test_compile(r"let x = 5 in \y: Int -> x");
}

#[test]
fn compile_app() {
    test_compile_and_execute(
        r"
let x = 5,
    f = \y -> x
in f 555
",
        5,
    );
}
