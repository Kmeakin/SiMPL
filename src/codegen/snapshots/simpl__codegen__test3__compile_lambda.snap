---
source: src/codegen/test3.rs
expression: module.print_to_string().to_string()
---
; ModuleID = 'test_compile'
source_filename = "test_compile"

%Closure = type { i8*, i8* }
%Env = type { i64 }

define %Closure @toplevel() {
toplevel_entry:
  %x = alloca i64
  store i64 5, i64* %x
  %closure = alloca %Closure
  %closure.code = getelementptr inbounds %Closure, %Closure* %closure, i32 0, i32 0
  store i8* bitcast (i64 (%Env, i64)* @lambda to i8*), i8** %closure.code
  %closure.env = getelementptr inbounds %Closure, %Closure* %closure, i32 0, i32 1
  %malloccall = tail call i8* @malloc(i32 ptrtoint (i64* getelementptr (i64, i64* null, i32 1) to i32))
  %closure.env1 = bitcast i8* %malloccall to %Env*
  %env.x = getelementptr inbounds %Env, %Env* %closure.env1, i32 0, i32 0
  %x2 = load i64, i64* %x
  store i64 %x2, i64* %env.x
  %closure.env3 = bitcast %Env* %closure.env1 to i8*
  store i8* %closure.env3, i8** %closure.env
  %closure4 = load %Closure, %Closure* %closure
  ret %Closure %closure4
}

define i64 @lambda(%Env %env, i64 %y) {
lambda_entry:
  %env1 = alloca %Env
  store %Env %env, %Env* %env1
  %env.x = alloca i64
  %env.x2 = getelementptr inbounds %Env, %Env* %env1, i32 0, i32 0
  %env.x3 = load i64, i64* %env.x2
  store i64 %env.x3, i64* %env.x
  %y4 = alloca i64
  store i64 %y, i64* %y4
  %x = load i64, i64* %env.x
  ret i64 %x
}

declare noalias i8* @malloc(i32)

